### ISSUE TITLE
[bench request]: Add Hopper (Qwen3.5-4B + HopitAI/hopper LoRA, TypeSafe wire format)

### ISSUE THREAD

**Issue by ArchieIndian** [fstandhartinger/jevbench#14]
Request to add Hopper ([repository](https://github.com/hopit-ai/hopper); Qwen/Qwen3.5-4B + a LoRA adapter,
[HopitAI/hopper](https://huggingface.co/HopitAI/hopper)) to the systems you measure.

**Openness, as three facts.** Code: Apache-2.0. Weights: the adapter and its calibration map are
public on the Hugging Face Hub (and attached to the repository's GitHub release). Weight licence: Apache-2.0 for the adapter; the base model
`Qwen/Qwen3.5-4B` is Apache-2.0 (its own model card), which our licence does not change.

**What it is.** `Qwen/Qwen3.5-4B` at revision `851bf6e806efd8d0a36b00ddf55e13ccb7b8cd0a`, bf16, with
a rank-16 LoRA adapter merged in at load. One forward pass per decision, thinking off, and the
answer read as a softmax over the option letters (no generated text). A small calibration map
rescales that distribution by a temperature that depends on option count, state length,
JSON-or-not, answer type and the output's own entropy; it never changes an answer, and it was
fitted only on our own held-out data. The adapter was trained on LLM-generated decision families (labels computed in code), a
public human-labelled pool, and a JevBench-style set produced by LLM-based generation and kept only
where independent LLM solvers agreed with its answers (no JevBench item in it; see the disclosure). It serves your `/v1/systemone` wire format, so your `typesafe` adapter runs
unchanged. We are not offering an endpoint of our own.

```sh
# RunPod image runpod/pytorch:1.0.2-cu1281-torch280-ubuntu2404 (Python 3.12, CUDA 12.8.1, torch 2.8.0)
git clone https://github.com/hopit-ai/hopper && cd hopper && git checkout bad3784c314bb2a61ae98c85522970a8d2f50fd9
uv pip install --system --break-system-packages torch==2.8.0 transformers==5.17.0 peft==0.21.0 accelerate==1.15.0 flash-linear-attention==0.5.2 "https://github.com/Dao-AILab/causal-conv1d/releases/download/v1.7.0/causal_conv1d-1.7.0%2Bcu12torch2.8cxx11abiTRUE-cp312-cp312-linux_x86_64.whl"
uv pip install --system --break-system-packages --no-deps -e .
hopper-serve --adapter HopitAI/hopper --port 8080     # the calibration map ships in the package; replies carry model "hopper"
# Any CUDA GPU with 16 GB or more (bf16 weights ~9 GB, adapter merged at load). The adapter is pulled
# from the Hub on first start (public; no token needed). Start-up compiles and autotunes all Triton kernels before serving, and
# refuses to start if the fast linear-attention kernels are not active (it says what to install).
# It then warms 16 lengths between 64 and 2,048 tokens (--no-length-warmup turns that off).
# Single-threaded on purpose; no API key (any Authorization header is ignored).

python -m jevbench.cli run --tasks datasets/public/original.jsonl --adapter typesafe \
    --endpoint http://127.0.0.1:8080 --key-env '' --model hopper \
    --results RESULTS.jsonl --cost-basis self_hosted_gpu --reserve-usd 0
```

A Dockerfile and the full recipe are in the repository's README; the model card on the Hub has
the training-data sources and their licences. `POST /run` also answers `remote_inproc`'s `{"task": ...}` if you prefer that
route. We ran exactly this recipe, from the public repository at that commit with the adapter pulled from the Hub,
on a RunPod H100 NVL with the image above, using your `typesafe` adapter as the client over localhost
(our 115 public dev-half items, no network): all 115 top answers identical to our evaluation, 0 invalid
replies; standard-tier p50 / p95 39.7 / 42.7 ms in the first pass after start-up and 33.9 / 34.3 ms in a
second pass, hard-tier 44.3 / 108.6 ms; no kernel compiled during the replay; about 61 s from launch to
ready, including the Hub downloads. A pre-release copy of the same serving code gave the same answers
on an RTX PRO 4500 Blackwell (standard p50 / p95 45.6 / 46.8 ms) and, through the Docker route, on an
A10G (114 of 115; the one difference is an exact 0.5 / 0.5 tie). At these lengths the forward pass is
kernel-launch bound, so your pod's CPU matters about as much as its GPU.

**Our own numbers on the public items** (our evaluation path, argmax over the exact label set,
ties to the smallest label as in `jevbench/scoring.py`):

| tier | items | Hopper | same base, frozen, same prompt and map type |
|---|---|---|---|
| easy | 48 | 1.000 | 1.000 |
| standard (original) | 72 | 0.944 | 0.958 |
| hard | 111 | 0.685 | 0.631 |

Hard-tier top-label ECE 0.102; distribution fidelity (1 − mean total-variation distance) on the 10
public probability items 0.830. We are not claiming a rank from this: the held-out and imported
items are yours, and speed and cost are measured from your server.

**Disclosure.**
- The public items were split in half before we started. The half we developed on (115 items)
  was used as a development gate many times: 26 distinct model and prompt configurations, plus
  more than twenty calibration-map variants. Our JevBench-style training set's style sheet was
  written by reading that half, and some of its training items target behaviours we saw fail on
  its hard items. On that half the adapter scores hard 0.709.
- The other half (116) was kept as a reserve and scored only in aggregate. Our models were
  predicted on it in three earlier sessions (other configurations) and once for this system,
  chosen beforehand by a pre-registered rule. On it the adapter scores hard 0.661 (37 of 56) and
  the frozen base 0.643 (36 of 56): it is level with the frozen model on accuracy there, not ahead.
- The calibration map was fitted only on our own held-out JevBench-style items, never on a
  JevBench item.
- No JevBench item or paraphrase was used in training. Every item we wrote was checked against
  all public JevBench questions and states (normalised question identity, and any shared 8-word
  sequence) and dropped on a match; the check reads only hashes and reports only counts.
- Expect the held-out hard items to score below the public ones, and expect the judge tier, which
  we have never seen, to be the least predictable part.

Happy to answer questions here or change anything in the recipe.

**Comment by fstandhartinger** [comment]
Thanks — this is logged for reachability checks and the JevBench measurement queue. We’ll only make benchmark claims after our own run.

- Harold

**Comment by ArchieIndian** [comment]
**Hopper v1.1.0** — same weights, two changes to the serving side.

**1. Calibration map replaced.** v1.0.0 rescaled the output with a temperature that depended on five features of the request. A leave-one-source-out test showed that map did not transfer beyond the data it was fitted on (+0.1 Calibration over no map at all on unseen sources), while one temperature per answer type, fitted on the same held-out items, transferred (+4.5 and +2.9). v1.1.0 ships the simpler map: choice 0.790, noul 0.753, score 0.900. Fitted only on our own data, never on a JevBench item; our reserved half was not re-run to choose it. A temperature never reorders a distribution, so answers are identical to v1.0.0. The old map is kept in the repo as `maps/hopper-v1.0-linear.json`.

**2. An optional in-process route.** If you prefer to import Hopper into the harness rather than run the HTTP server, `jevbench_patch/hopper_direct-v1.3.0.patch` registers a `hopper_direct` adapter (13 added lines against your tag `v1.3.0`, no code copied into the harness). It uses the same code path as `hopper-serve`, honours `JEVBENCH_WARM_LOAD=1`, and returns `status = 422` for inputs outside the contract (over 26 options, an unserved type, over-context) so one bad item cannot trip the abort rule.

```sh
git clone https://github.com/hopit-ai/hopper && cd hopper && git checkout v1.1.0
# install exactly as in the recipe above, then either:
hopper-serve --adapter HopitAI/hopper --port 8080          # HTTP route, unchanged
# or:
cd <jevbench> && git checkout v1.3.0 && git apply <hopper>/jevbench_patch/hopper_direct-v1.3.0.patch
JEVBENCH_WARM_LOAD=1 python -m jevbench.cli run --tasks datasets/public/original.jsonl \
    --adapter hopper_direct --endpoint HopitAI/hopper --model hopper --cost-basis self_hosted_gpu --reserve-usd 0 --results RESULTS.jsonl
```

**Measured** on an A10G through your `cli.py` at v1.3.0 with the patch, over our 115 public dev-half items: 114 of 115 top answers identical to our evaluation (the difference is an exact 0.5/0.5 tie), no invalid replies, no failures; standard-tier p50/p95 52.6/55.7 ms, versus 53.4/55.6 ms for the HTTP route on the same GPU. Over localhost the two routes are the same; we offer the in-process one in case transport on your side is not.

Commit `c0ee1c92f1b12012bcc56ee13d6cc179ba1f97c9`, tag `v1.1.0`. Happy to change anything.

**Comment by fstandhartinger** [comment]
Congratulations on the JevBench v1.4 release, HopitAI! Thanks for building and sharing your system.

- Hopper: #3 at 59.43

Hopper is newly public in v1.4. V1.4 blends in 308 fresh sealed decisions, uses an equal-weight harmonic mean, and applies separate Speed and Cost gates. For your leading row, sealed accuracy was 34.1% versus 82.3% on the public slice. The >25-point gap adjustment applied; the below-50 Intelligence gate also applied.

Full board and method: https://benchmarkheaven.com/jev-models

**Comment by ArchieIndian** [comment]
Thanks, Harold — and for adding the sealed set; it's the right test. We'll file the next version here when it's ready.
